benchmark/main.cpp

```diff
@@ -56,979 +50,1198 @@
+            goto exit;
+        }
+
+        memset(buf, 0, sizeof(buf));
+        memset(tmp, 0, sizeof(tmp));
+
+        if ((ret = mbedtls_aes_setkey_enc(&aes, tmp, keysize)) != 0) {
+            mbedtls_printf("mbedtls_aes_setkey_enc() returned -0x%04X\n",
+                           -ret);
+            goto exit;
         }
-        mbedtls_aes_free( &aes );
+
+        BENCHMARK_FUNC_CALL(title,
+                            mbedtls_aes_crypt_cbc(&aes,
+                                                  MBEDTLS_AES_ENCRYPT, BUFSIZE,
+                                                  tmp, buf, buf));
     }
-#endif
 
-#if defined(MBEDTLS_CIPHER_MODE_CTR)
-    if( todo->aes_ctr )
-    {
-        int keysize;
-        size_t nc_offset = 0;
-        unsigned char stream_block[16];
-        mbedtls_aes_context aes;
-        mbedtls_aes_init( &aes );
-        for( keysize = 128; keysize <= 256; keysize += 64 )
-        {
-            mbedtls_snprintf( title, sizeof( title ), "AES-CTR-%d", keysize );
-
-            memset( buf, 0, sizeof( buf ) );
-            memset( tmp, 0, sizeof( tmp ) );
-            mbedtls_aes_setkey_enc( &aes, tmp, keysize );
-
-            TIME_AND_TSC( title,
-                mbedtls_aes_crypt_ctr( &aes, BUFSIZE, &nc_offset, tmp, stream_block, buf, buf ) );
+exit:
+    mbedtls_aes_free(&aes);
+#endif /* MBEDTLS_AES_C && MBEDTLS_CIPHER_MODE_CBC */
+
+    return ret;
+}
+
+MBED_NOINLINE static int benchmark_aes_ctr() {
+    int ret = 0;
+#if defined(MBEDTLS_AES_C) && defined(MBEDTLS_CIPHER_MODE_CTR)
+    int keysize;
+    size_t nc_offset = 0;
+    unsigned char stream_block[16];
+    mbedtls_aes_context aes;
+
+    mbedtls_aes_init(&aes);
+
+    for ( keysize = 128; keysize <= 256; keysize += 64 ) {
+        ret = mbedtls_snprintf(title, sizeof(title), "AES-CTR-%d", keysize);
+        if (ret < 0 || static_cast<size_t>(ret) >= sizeof(title)) {
+            mbedtls_printf("Failed to compose title string using "
+                           "mbedtls_snprintf(): %d\n", ret);
+            goto exit;
         }
-        mbedtls_aes_free( &aes );
+
+        memset(buf, 0, sizeof(buf));
+        memset(tmp, 0, sizeof(tmp));
+
+        if ((ret = mbedtls_aes_setkey_enc(&aes, tmp, keysize)) != 0) {
+            mbedtls_printf("mbedtls_aes_setkey_enc() returned -0x%04X\n",
+                           -ret);
+            goto exit;
+        }
+
+        BENCHMARK_FUNC_CALL(title,
+                            mbedtls_aes_crypt_ctr(&aes, BUFSIZE, &nc_offset,
+                                                  tmp, stream_block, buf,
+                                                  buf));
     }
-#endif
 
-#if defined(MBEDTLS_GCM_C)
-    if( todo->aes_gcm )
-    {
-        int keysize;
-        mbedtls_gcm_context gcm;
+exit:
+    mbedtls_aes_free(&aes);
+#endif /* MBEDTLS_AES_C && MBEDTLS_CIPHER_MODE_CTR */
+
+    return ret;
+}
+
+MBED_NOINLINE static int benchmark_aes_gcm() {
```

Comment:
why not surround the whole function with `#if defined(MBEDTLS_AES_C) && defined(MBEDTLS_GCM_C)`?